### PR TITLE
More efficient way of computing skip slot cache key

### LINKED ISSUE #11433: Don't use HTR for the cache key in ssc
We currently use the beacon-block header HTR hashed against the slot for the skip slot cache key. This requires two calls to hashing routines that are completely unnecessary in in 
```go
func cacheKey(_ context.Context, state state.ReadOnlyBeaconState) ([32]byte, error) {
	bh := state.LatestBlockHeader()
	if bh == nil {
		return [32]byte{}, errors.New("block head in state can't be nil")
	}
	r, err := bh.HashTreeRoot()
	if err != nil {
		return [32]byte{}, err
	}
	return hash.Hash(append(bytesutil.Bytes32(uint64(state.Slot())), r[:]...)), nil
}
```
We can simply use the `StateRoot` in `bh` instead of `bh.HashTreeRoot()` and we can just replace the last 8 bytes with the slot instead of hashing.

### PR DESCRIPTION
**Pull request by terencechain**
**What does this PR do? Why is it needed?**
This PR implements a more efficient way of computing for skip slot cache keys. This current mechanism to compute for skip slot cache requires two hashes which are unnecessary and 30x slower by the benchmark. I explored just plugging in state root and keeping `Hash` and that's 7x slower. 
```
Current implementation:
BenchmarkCacheKey-12    	  380469	      3261 ns/op

Keep `Hash` but eliminate `HTR`:
BenchmarkCacheKey-12    	 1756923	       649.1 ns/op

This PR:
BenchmarkCacheKey-12    	 6763269	       174.7 ns/op
```

I had to fix a few bad tests that were incorrectly dependent on skip slot cache to pass. The mock for HeadState should return the copy instead of same reference to keep using mutated state

**Which issues(s) does this PR fix?**

Fixes #11433

**Other notes for review**
I exposed `cacheKey` -> `SkipSlotCacheKey` for tests. I'm open to suggestions on how to best deal with this

### CODE REVIEW

**Comment by rkapka** [comment]
> I exposed cacheKey -> SkipSlotCacheKey for tests. I'm open to suggestions on how to best deal with this

How about moving cache key tests to a new file with `package transition`?

**Comment by terencechain** [comment]
> I see the existing test file has `package transition_test`. How about moving cache key tests to a new file with `package transition`?

Harder than I thought, getting import cycle errors now 

```
FAIL	github.com/prysmaticlabs/prysm/v3/beacon-chain/core/transition [setup failed]
package github.com/prysmaticlabs/prysm/v3/beacon-chain/core/transition
	imports github.com/prysmaticlabs/prysm/v3/testing/util: import cycle not allowed in test
```

File: beacon-chain/core/transition/skip_slot_cache.go [hunk 1]
```diff
@@ -16,17 +15,19 @@ import (
...
 	bh := state.LatestBlockHeader()
 	if bh == nil {
 		return [32]byte{}, errors.New("block head in state can't be nil")
 	}
-	r, err := bh.HashTreeRoot()
-	if err != nil {
-		return [32]byte{}, err
-	}
-	return hash.Hash(append(bytesutil.Bytes32(uint64(state.Slot())), r[:]...)), nil
+	sr := bh.StateRoot
```
Comment (prestonvanloon) [inline beacon-chain/core/transition/skip_slot_cache.go:28]: What do you think about a defensive check here?

```suggestion
	sr := bh.StateRoot
	if len(sr) != 32 {
		return [32]byte{}, errors.New("invalid state root in latest block header")
	}
```

**Comment by prestonvanloon** [comment]
Many tests fail

File: beacon-chain/core/transition/skip_slot_cache_test.go [hunk 2]
```diff
@@ -80,6 +82,8 @@ func TestSkipSlotCache_ConcurrentMixup(t *testing.T) {
 		require.NoError(t, err)
 		s1, err = transition.ExecuteStateTransition(context.Background(), originalState.Copy(), wsb)
 		require.NoError(t, err, "Could not run state transition")
+		s1, err = transition.ProcessSlot(context.Background(), s1)
```
Comment (terencechain) [inline beacon-chain/core/transition/skip_slot_cache_test.go:85]: This current test setup is unrealistic in the production environment. The state root is zeros until we call `ProcessSlot`, this only applies to genesis condition

**Comment by rauljordan** [comment]
hi @terencechain PTAL at the failing tests. They seem like they shouldnt be failing

File: beacon-chain/blockchain/testing/mock.go [hunk 3]
```diff
@@ -281,7 +281,10 @@ func (s *ChainService) HeadBlock(context.Context) (interfaces.SignedBeaconBlock,
 
 // HeadState mocks HeadState method in chain service.
 func (s *ChainService) HeadState(context.Context) (state.BeaconState, error) {
-	return s.State, nil
+	if s.State == nil {
+		return nil, nil
+	}
+	return s.State.Copy(), nil
```
Comment (terencechain) [inline beacon-chain/blockchain/testing/mock.go:287]: We never want to mutate the mock state

File: beacon-chain/rpc/eth/validator/validator_test.go [hunk 4]
```diff
@@ -176,6 +176,9 @@ func TestGetAttesterDuties(t *testing.T) {
 		}
 		resp, err := vs.GetAttesterDuties(ctx, req)
 		require.NoError(t, err)
+
+		bs, err = transition.ProcessSlotsIfPossible(ctx, bs, params.BeaconConfig().SlotsPerEpoch*2)
```
Comment (terencechain) [inline beacon-chain/rpc/eth/validator/validator_test.go:180]: Fixed a bad testsetup. This was relied on mutated mock state

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix_test.go [hunk 5]
```diff
@@ -528,6 +529,7 @@ func TestServer_getAndBuildHeaderBlock(t *testing.T) {
 }
 
 func TestServer_GetBellatrixBeaconBlock_HappyCase(t *testing.T) {
+	transition.SkipSlotCache.Disable()
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_bellatrix_test.go:532]: Cached skip slot state was leaking

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_test.go [hunk 6]
```diff
@@ -2089,6 +2090,7 @@ func TestProposer_GetBeaconBlock_PostForkEpoch(t *testing.T) {
 }
 
 func TestProposer_GetBeaconBlock_BellatrixEpoch(t *testing.T) {
+	transition.SkipSlotCache.Disable()
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_test.go:2093]: Cached skip slot state was leaking

**Comment by terencechain** [comment]
> hi @terencechain PTAL at the failing tests. They seem like they shouldnt be failing

Ready again

**Review by rauljordan — APPROVED** [review]

**Pull request merged**